types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,711 @@
+	request->n_connectors = n_connectors;
+}
+
+static void drm_lease_request_v1_handle_submit(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *lease_resource = wl_resource_create(client,
+			&wp_drm_lease_v1_interface, version, id);
+	if (!lease_resource) {
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	struct wlr_drm_lease_request_v1 *request =
+			drm_lease_request_v1_from_resource(resource);
+	if (!request) {
+		wlr_log(WLR_ERROR, "Request has been destroyed");
+		wp_drm_lease_v1_send_finished(lease_resource);
+		return;
+	}
+
+	for (size_t i = 0; i < request->n_connectors; ++i) {
+		struct wlr_drm_lease_connector_v1 *conn = request->connectors[i];
+		if (conn->active_lease) {
+			wlr_log(WLR_ERROR, "Failed to create lease, connector %s has "
+					"already been leased", conn->output->name);
+			wp_drm_lease_v1_send_finished(lease_resource);
+			return;
+		}
+	}
+
+	struct wlr_drm_lease_v1 *lease = calloc(1, sizeof(struct wlr_drm_lease_v1));
+	if (!lease) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_v1");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	lease->device = request->device;
+	lease->resource = lease_resource;
+	wl_list_insert(&lease->device->leases, &lease->link);
+
+	wl_resource_set_implementation(lease_resource, &lease_impl, lease,
+			drm_lease_v1_handle_resource_destroy);
+
+	request->lease = lease;
+
+	if (request->invalid) {
+		/* Pre-emptively reject invalid lease requests */
+		wlr_log(WLR_ERROR, "Invalid request");
+		drm_lease_v1_destroy(lease);
+	} else {
+		/* TODO: reject the request if the user does not grant it */
+		wlr_signal_emit_safe(&request->device->manager->events.request,
+				request);
+	}
+
+	/* Request is done */
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_request_v1_interface lease_request_impl = {
+	.request_connector = drm_lease_request_v1_handle_request_connector,
+	.submit = drm_lease_request_v1_handle_submit,
+};
+
+static void drm_lease_device_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void drm_lease_device_v1_handle_release(
+		struct wl_client *client, struct wl_resource *resource) {
+	wp_drm_lease_device_v1_send_released(resource);
+	wl_resource_destroy(resource);
+}
+
+static void drm_lease_device_v1_handle_create_lease_request(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *request_resource = wl_resource_create(client,
+		&wp_drm_lease_request_v1_interface, version, id);
+	if (!request_resource) {
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		return;
+	}
+
+	wl_resource_set_implementation(request_resource, &lease_request_impl,
+		NULL, drm_lease_request_v1_handle_resource_destroy);
```

Comment:
We can move this below and replace the `wl_resource_set_user_data` call.

Reply:
As discussed on IRC, a `wl_resource` needs an implementation in case a request on an inert object is made.